More HAL refactoring

martukas opened the pull request:
This all but eliminates global variables and their static initialization, other than:
* `god_object` in `main.cpp`
* `hal` as part of `HalApi`

Most significantly, `v03system.*` fully encapsulates all choices of classes and channel mappings that may depend on PCB and processor architecture. There are now abstract classes for the Sensors, Actuators and System containers. The latter might be slowly becoming the new HAL, conceptually.

Furthermore, all analog sensor implementations are now template classes that may have any mock object injected for testing their logic. A number of other classes, such as `Debug::Interface` also use more `std::optional` to defer initialization.

Doxygen "TODO" syntax (at least partially) harmonized.

Code has been regression tested with:
* deploying
* using debug interface with `get all` 
* using debug interface with `test run dummy` to verify breathing and data acquisition
* running with GUI and changing RR to verify comms interface and behavior

What's next?:
* move some of the code from `MainLoop` to `V03System`
* make `MainLoop` a template class, so it can be tested with mock HALs
* refactor stepper-related code at least enough to get it out of `hal_stm32.cpp`
* make `hal_stm32` mostly about interrupts, or maybe even move that out into separate module, e.g. may be that `early_init()` is not too tightly coupled to it
* make it so that whatever is left of the old "HAL" is truly abstract and initialized in scope, so we can have pretty much only one "ifdef" block in a single place to selectively include everything to do with v0.3 configuration

Updates #1184
Updates #1262 
Updates #1146 

## Self-review checklist:

- [x] Self-review: looked through the `Files changed` tab, browsed repository in branch
- [x] Documentation updated - reflects changes in code, electrical or mechanical design
- [x] Documentation and graphics follow the [documentation style guide](https://github.com/RespiraWorks/Ventilator/wiki/Documentation-style-guide)
- [x] New content is linked, easily discoverable, does not require too many clicks
- [x] Follows other relevant parts of [contributor wiki](https://github.com/RespiraWorks/Ventilator/wiki)
- [x] PR has a descriptive name
- [x] Tagged relevant reviewers

### Software only:

- [x] Follows our [code style](https://github.com/RespiraWorks/Ventilator/wiki/Code-style)
- [x] Commented code, particularly in hard-to-understand areas
- [x] All tests pass - new and old, locally and on CI
- [x] No new warnings or static check failures
- [x] Tests that prove fix is effective or new feature works
- [x] Manual tests are explained, with instructions for reproducing them

asmodai27 commented on software/controller/lib/sensors/analog_sensor.tpp:
```diff
@@ -13,11 +13,10 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
```
Maybe add warning not to include, similar to other tpp files ?

martukas replied:
Done.

asmodai27 commented on software/controller/lib/sensors/oxygen.tpp:
```diff
@@ -13,25 +13,25 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
```
Again, maybe add a warning not to include outside of corresponding header?

asmodai27 commented on software/controller/lib/sensors/pressure_sensors.tpp:
```diff
@@ -13,17 +13,17 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
```
same here

asmodai27 commented on software/controller/src/main.cpp:
```diff
@@ -13,220 +13,49 @@ See the License for the specific language governing permissions and
-  return proto;
-}
+/// \TODO may not need this if high priority functor could be bound
+static MainContainer god_object;
```
Not sure about that name, though I get the idea.

asmodai27 reviewed (commented):
I like this. Here are some minor comments on the code, pending testing on my unit to approve the PR for good.

asmodai27 reviewed (commented):
A few more comments after testing, and getting more familiar with this code.

martukas commented on software/controller/lib/core/comms.h:
```diff
@@ -38,6 +41,7 @@ class Comms {
 
  private:
+  /// \TODO does this really need to be here? Not directly used
   UART::Channel *uart_;
```
OK, removing it in this PR then.

martukas commented on software/controller/integration_tests/eeprom_test.h:
```diff
@@ -82,14 +87,19 @@ static Debug::Command::VarHandler var_command;
+  debug.add_handler(Debug::Command::Code::Variable, &var_command);
+  debug.add_handler(Debug::Command::Code::Trace, &trace_command);
+  debug.add_handler(Debug::Command::Code::EepromAccess, &eeprom_command);
+
```
Maybe I won't touch these for now, because I see that this integration test could use even more work. Let's leave the TODO and take care of it in another PR.

martukas commented on software/controller/lib/core/main_loops.h:
```diff
@@ -0,0 +1,78 @@
+  void background_task();
+  static SensorsProto AsSensorsProto(const SensorReadings &r, const ControllerState &c);
+
+  System hardware_layer;
```
Indeed. I was going leave that for later, but might as well clean it up now.

martukas commented on software/controller/lib/hal/gpio.h:
```diff
@@ -36,7 +36,6 @@ Reference abbreviations [RM], [DS], etc are defined in hal/README.md.
 namespace GPIO {
 
-/// \TODO: Move these mappings elsewhere
 enum class Port : uint32_t {
```
I will add a TODO, but not touch this for now

martukas commented on software/controller/lib/hal/gpio.h:
```diff
@@ -46,23 +45,7 @@ enum class Port : uint32_t {
-
+/// \TODO make these more generic?
 // [DS] Tables 17-18 (pg 76-86)
 enum class AlternativeFunction : uint32_t {
```
ditto

martukas commented on software/controller/lib/hal/gpio_mock.cpp:
```diff
@@ -21,13 +21,11 @@ Reference abbreviations [RM], [DS], etc are defined in hal/README.md.
 */
 #if !defined(BARE_STM32)
```
Yes, but the problem is that we are still using ifdefs instead of implementing an API defined in abstract class. This will require further work to remove this. I have added a TODO into gpio_stm32.cpp about this and leaving this next step for another PR

martukas commented on software/controller/lib/core/v03system.cpp:
```diff
@@ -0,0 +1,180 @@
+}
+
+V03Actuators::V03Actuators()
+    : blower_pinch_("blower_valve_", " of the blower pinch valve", BlowerValveMotorIndex),
```
Actually, I made both pinch valves `std::optional` now and they initialize in `init()`.
We want to keep all this in the cpp, because I'd rather include `system_constants.h` here and leave the definition as generic as possible.

martukas commented on software/controller/lib/core/v03system.cpp:
```diff
@@ -0,0 +1,180 @@
+
+void V03Actuators::beep(float volume) { buzzer_->set(volume); }
+
+V03System::V03System()
```
Nope. See above.

martukas commented on software/controller/lib/core/v03system.cpp:
```diff
@@ -0,0 +1,180 @@
+}
+
+void V03Actuators::execute(const ActuatorsState &desired_state) {
+  // set blower PWM
```
I don't think this check is necessary every time? I don't think there is a way for them to become not ready after they have become ready when initialized/homed/etc..? I added a comment about the assumption, which later we can upgrade to be a `\precondition` according to Doxygen syntax whenever we have that party...

asmodai27 replied:
Maybe add a more verbose version of that comment in the header file, so the caller knows about it then?

martukas commented:
> Tests I ran were OK for most use cases, but:
> 
> * `stepper` and `pinch valve` integration tests are broken (tested OK on `master`)
> * `eeprom` integration test is broken (tested OK on `master`, though I cannot use the debug interface as is, due to the debug scripts looking for `0_controller_version` and other variables at startup).

Fixed unit tests and modified the debug interface to be more tolerant of missing variables. It will still complain about it, but should not crash.
Eeprom unit test definitely needed the most modification, but partly its because of the presence of the debug interface. I think It might make sense to have a more minimal eeprom test just for confirming the subsystem works with minimal human intervention. A minimal controller version that has quite some sophisticated debug facilities is maybe in a different category - we still need it, but I think it could be a separate thing. I added a TODO about that.

Actually, I was not even aware of some additional procedure where I'd have to be checking something via debug interface. If that is a test we are expected to perform, we should add it to some "manual regression testing" checklist.

martukas commented on software/controller/lib/core/main_loops.cpp:
```diff
@@ -0,0 +1,167 @@
+  SensorsProto proto = SensorsProto_init_zero;
+  proto.patient_pressure_cm_h2o = r.patient_pressure.cmH2O();
+  proto.inflow_pressure_diff_cm_h2o = 0;   // \TODO field unused and obsolete, should change proto
+  proto.outflow_pressure_diff_cm_h2o = 0;  // \TODO field unused and obsolete, should change proto
```
Fixed

asmodai27 reviewed (approved):
Looks good now.
Some minor notes here and there.

martukas commented on software/controller/lib/core/main_loops.cpp:
```diff
@@ -141,9 +103,9 @@ void MainContainer::background_task() {
   // Override received gui_status from the RPi with values from DebugVars iff the forced_mode
   // DebugVar has a legal value.
-  if (uint32_t m = forced_mode.get(); m >= _VentMode_MIN && m <= _VentMode_MAX) {
+  if (uint32_t mode = forced_mode.get(); mode >= _VentMode_MIN && mode <= _VentMode_MAX) {
```
The way to give control back to GUI is
```
set forced_mode invalid
```
which the debug interface accepts, and then the controller ignores the forced mode.

The question might be whether we want to add a formal "invalid" mode to the protocol definition?
